spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/SimpleRedBlackUpgradeStrategy.java

```diff
@@ -108,50 +109,16 @@ public Release upgrade(Release existingRelease, Release replacingRelease,
 		appDeployerData.setDeploymentDataUsingMap(appNameDeploymentIdMap);
 		this.appDeployerDataRepository.save(appDeployerData);
 
-		// TODO update status in DB that the other apps are being deleted.
-
-		this.delete(existingRelease, applicationNamesToUpgrade);
+		AppDeployerData replacingAppDeployerData = this.appDeployerDataRepository
+				.findByReleaseNameAndReleaseVersionRequired(
+						replacingRelease.getName(), replacingRelease.getVersion());
 
-		// Update Status in DB
-		Status status = new Status();
-		status.setStatusCode(StatusCode.DEPLOYED);
-		replacingRelease.getInfo().setStatus(status);
-		replacingRelease.getInfo().setDescription("Upgrade complete");
+		// Executes on another thread.
+		// TODO Investigate why lookups in appDeployerDataRepository do not return expected results
+		this.healthCheckAndDeleteStep.waitForNewAppsToDeploy(existingRelease, existingAppDeployerData,
```

Comment:
Don't understand why this doesn't error in spring as I'd assume spring `@Async` should use `CompletableFuture` but this function just returns null. Maybe error is discarded or spring checks that actual returned value is null. Anyway this is somewhat wrong.

Reply:
Also I'm not sure if whole `Transactional` stuff work anymore with this.

Reply:
Will look to split the logic into two transactions, such that there won't be two concurrent transactions occuring in the same request lifecycle.